File: 0605-can-place-flowers/0605-can-place-flowers.java
class Solution {
    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        int plantedFlowers = 0;
        boolean plantedAllFlowers = false;
        for (int i = 0; i < flowerbed.length; i++) {
//          Check if the current plot is empty.
            if (flowerbed[i] == 0) {
//                  check to see if we are at 0 or if the plot to the left is 0
                boolean emptyLeftPlot = (i == 0) || (flowerbed[i - 1] == 0);
//                  checks to see if we are at the end of the flowerbed or if the plot to the right is 0
                boolean emptyRightPlot = (i == flowerbed.length - 1) || (flowerbed[i + 1] == 0);
                
//                  if both plots are empty (0), then we can plant
                if (emptyLeftPlot && emptyRightPlot) {
//                     set the flowerbed at our current index to 1 ie we just planted
                    flowerbed[i] = 1;
//                     increment the number of planted flowers
                    plantedFlowers++;
                }
            }
        }
//         if we planted n or more flowers, then we were able to plant all n flowers with no isses
        if (plantedFlowers >= n) {
            plantedAllFlowers = true;
        }
        
        return plantedAllFlowers;
    }
}

File: 0605-can-place-flowers/0605-can-place-flowers.js
/**
 * @param {number[]} flowerbed
 * @param {number} n
 * @return {boolean}
 */
var canPlaceFlowers = function(flowerbed, n) {
    var plantedFlowers = 0;
    for (let i = 0; i < flowerbed.length; i++) {
        if (flowerbed[i] === 0) {
            var emptyLeft = ((i === 0) || (flowerbed[i - 1] === 0));
            var emptyRight = ((i === flowerbed.length - 1) || (flowerbed[i + 1] === 0));
            
            if (emptyLeft && emptyRight) {
                flowerbed[i] = 1;
                plantedFlowers++;
            }
        }
    }
    return plantedFlowers >= n;
};

// var canPlaceFlowers = function(flowerbed, n) {
//     flowerbed.unshift(0); // Add a boundary element at the beginning
//     flowerbed.push(0); // Add a boundary element at the end
    
//     var plantedFlowers = 0;
//     for (let i = 1; i < flowerbed.length - 1; i += 2) {
//         if (flowerbed[i] === 0 && flowerbed[i - 1] === 0 && flowerbed[i + 1] === 0) {
//             flowerbed[i] = 1;
//             plantedFlowers++;
//         }
//         if (plantedFlowers >= n) {
//             return true; // Early exit if the required number of flowers has been planted
//         }
//     }
    
//     return plantedFlowers >= n;
// };